### ISSUE TITLE
ector loses all brain when restarted

### ISSUE THREAD

**Issue by victorperin** [parmentf/hubot-ector#1]
When I restart hubot, it loses all conversation.
Is there a way to save just the most  talked thing to disk?

**Comment by victorperin** [comment]
Is there a way to use it with redis-brain.coffee?

Maybe changing:

``` coffeescript
  robot.catchAll (msg) ->
    if not quiet
      text = msg.message.text
      ector.setUser msg.message.user.name
      ector.addEntry text
      ector.linkNodesToLastSentence previousResponseNodes
      response = ector.generateResponse()
      previousResponseNodes = response.nodes
      msg.reply response.sentence
```

to something like:

``` coffeescript
  robot.catchAll (msg) ->
    if not quiet
      text = msg.message.text
      ector.setUser msg.message.user.name
      if not loaded_brain
        ector_brain = robot.brain.get('ector_brain') #need to be some type of table 
        ector.addEntry ector_brain
        loaded_brain = true
      else
        ector.addEntry text
        ector_brain = ector_brain+text #this line should insert the value of text inside ector_brain table.
      ector.linkNodesToLastSentence previousResponseNodes
      response = ector.generateResponse()
      previousResponseNodes = response.nodes
      msg.reply response.sentence
```

For now we need to figure out how get something like a table from robot.brain and then a way to put these values into ector.

**Comment by parmentf** [comment]
It's been a while since I wrote this.
Of course, you could use `robot.brain`, and, even, you can get inspired by this: https://github.com/parmentf/node-ector-command/blob/0.1.2/lib/ector_command.js#L149 to get the `cnJson`, and save it to the `robot.brain` (sorry, this is Javascript):

``` javascript
var cnJson = new Buffer(JSON.stringify(ector.cn));
```

`cnJson` will contain all the conversations, but also tokens, and links between them. 
You have to find a way to load it at start (and to save it on a regular basis).

Not much time these days to look in.

**Comment by parmentf** [comment]
Or take a look at this: https://github.com/parmentf/node-file-concept-network/blob/master/lib/file-concept-network.js

This is an example of how to save to a file, and to load from a file (really, it is a matter of ConceptNetwork).

**Comment by victorperin** [comment]
Oh, thanks.
I'm just getting sterted at hubot scripting and I loved this node!

Thanks, I'll try it.

**Comment by victorperin** [comment]
@parmentf sorry, but I'm confused here! hahaha
I choose to use your file-concept-network but I'm stuck in here:

``` coffeescript
var FileConceptNetwork = require('file-concept-network').FileConceptNetwork;
var fcn = new FileConceptNetwork();
var node1 = fcn.addNode("ECTOR");
var node2 = fcn.addNode("knows");
var node3 = fcn.addNode("Achille");
fcn.addLink(node1.id, node2.id);
fcn.addLink(node2.id, node3.id);
fcn.save('cn.json', function (err) {
  if (err) {
    console.error(err);
  }
});
```

So, how can I get these 3 nodes?
is it node1='Robot Name', node2='Message' and node3='User'?

**Comment by victorperin** [comment]
If is it, consider it done. And how can I put this data back in ector?

**Comment by parmentf** [comment]
Oh, sorry, I did not explain much.
You can look at https://github.com/parmentf/node-ector/blob/master/test/ector_test.js#L402-L445 to see how to inject a FileConceptNetwork into an Ector, instead of the in-memory default ConceptNetwork (just after Ector instanciation):

``` javascript
ector.injectConceptNetwork(FileConceptNetwork);
```

Then, to save, use: `ector.cn.save('cn.json')` (where `cn.json` can be replaced with a file path).
To load, use: `ector.cn.load('cn.json')` (or whatever file path you use to save).

**Comment by victorperin** [comment]
Ok, thanks again!
I'll make it! =)

**Comment by victorperin** [comment]
@parmentf I've tried here, but no success. =(
So, I was thinking about to make 2 commands: save the ConceptNetwork to a File and other to load this Network. This is diferent from saving the network every time.
So, from this npm (hubot-hector), is possible using file-concept-network as dependency and, when receive some command, copy the full concept network to a file? Maybe something like:

``` coffeescript
util = require 'util'
Ector = require 'ector'
FileConceptNetwork = require("file-concept-network").FileConceptNetwork
fcn = new FileConceptNetwork()

file_backup = "cn.json"
ector = new Ector()
previousResponseNodes = fcn.load file_backup
quiet = false

module.exports = (robot) ->

  ector.setName robot.name

  robot.catchAll (msg) ->
    if not quiet
      text = msg.message.text
      ector.setUser msg.message.user.name
      ector.addEntry text
      ector.linkNodesToLastSentence previousResponseNodes
      response = ector.generateResponse()
      previousResponseNodes = response.nodes
      msg.reply response.sentence

  robot.respond /save yourself/i, (msg) ->
    fcn.loadFromObject previousResponseNodes
    fcn.save file_backup
    msg.reply "Thanks, My mind is safe now!"
```

**Comment by victorperin** [comment]
I maked this function in file-concept-network but I can't do tests from here. (And I do not know these languages so much)

``` Javascript
FileConceptNetwork.prototype.loadFromObject = function (ob, cb) {
  var self = this;
    Object.merge(ob, JSON.parse(data));
    for (var key in ob) {
      self[key] = ob[key];
    }
    if (cb) { cb(err, self); }
};
```

**Comment by parmentf** [comment]
Leaving syntax problems apart, I think you still need (just after `new Ector()`) to inject the FileConceptNetwork into `ector`:

``` coffeescript
ector.injectConceptNetwork FileConceptNetwork
```

Then, you use `ector.cn` (which points to the ConceptNetwork instanciated) to load and save.

So, to load the backup (assuming `file_backup` contains the path to a file):

``` coffeescript
ector.cn.load file_backup
```

And to save to file_backup:

``` coffeescript
ector.cn.save file_backup
```

Oh: this is not a matter of `previousResponseNodes`, forget loading and saving from it (it contains roughly only the words from the last response generated by ector).

`ector.cn` is an instance of FileConceptNetwork, and contains the network.

**Comment by victorperin** [comment]
Ok, now it's giving this error:

```
Error: NewConceptNetwork is not a ConceptNetwork
  at Object.Ector.injectConceptNetwork 
```

My script are exactly this, what what else is wrong?

``` coffeescript
util = require 'util'
Ector = require 'ector'
FileConceptNetwork = require("file-concept-network").FileConceptNetwork
fcn = new FileConceptNetwork()

file_backup = "cn.json"
ector = new Ector()
ector.injectConceptNetwork FileConceptNetwork
previousResponseNodes = null
quiet = false

module.exports = (robot) -> ...
```

**Comment by parmentf** [comment]
Very strange. Maybe it has to do with CoffeeScript?
FileConceptNetwork is written as a derived class from ConceptNetwork.
This error should not arise.

**Comment by victorperin** [comment]
Ok I changed the way you told me and this occurs:

```
ERROR Unable to load src/ector: TypeError: object is not a function
  at Object.Ector.injectConceptNetwork (node_modules/ector/lib/ector.js:334:15)
  at Object.<anonymous> (src/ector.coffee:27:6, <js>:14:9)
  at Object.<anonymous> (src/ector.coffee:21:1, <js>:59:4)
  at Module._compile (module.js:456:26)
  at Object.loadFile (hubot/node_modules/coffee-script/lib/coffee-script/coffee-script.js:182:19)
  at Module.load (hubot/node_modules/coffee-script/lib/coffee-script/coffee-script.js:211:36)
  at Function.Module._load (module.js:312:12)
  at Module.require (module.js:364:17)
  at require (module.js:380:17)
  at Robot.loadFile (hubot/src/robot.coffee:216:9, <js>:162:11)
  at index.coffee:12:11, <js>:26:33
  at Object.cb [as oncomplete] (fs.js:168:19)
```

So, I'll convert the code into javascript and see what happens...

**Comment by victorperin** [comment]
The javascript:

``` javascript
var Ector, FileConceptNetwork, ector, fcn, file_backup, previousResponseNodes, quiet, speakReplies, util;
util = require('util');
Ector = require('ector');
FileConceptNetwork = require("file-concept-network").FileConceptNetwork;
file_backup = "cn.json";
ector = new Ector();
ector.injectConceptNetwork(FileConceptNetwork);
previousResponseNodes = null;
quiet = false;
module.exports = function(robot) {
...
};
```

**Comment by parmentf** [comment]
I'm trying to use FileConceptNetwork in ector-command (in JavaScript), and I get the same error message:

```
Error: NewConceptNetwork is not a ConceptNetwork
```

I'm puzzled!

**Comment by parmentf** [comment]
OK. I remember now: I never really used FileConceptNetwork (that's why ector-command writes it in its own code).

Maybe the right way is to do like in browser-ector (there, I'm sure it's working): https://github.com/parmentf/browser-ector/blob/master/js/ector.js#L60-L78

**Comment by parmentf** [comment]
Thanks to you, I made ector-command work using FileConceptNetwork! :)

Using ector@0.1.7 and file-concept-network@0.1.2, I do not have any error message (I hope I rightly published them all).

I succeeded in saving the ConceptNetwork. But I don't remember the syntax to pass a callback as a second argument in CoffeeScript, such that I got an error message while saving (though it was not a problem).

Here is what I wrote:

``` coffeescript
  robot.respond /save yourself/i, (msg) ->
    ector.cn.save file_backup
    msg.reply "Thanks, My mind is safe now!"
```

**Comment by parmentf** [comment]
To be complete, here is what works (with some error messages due to the lack of a callback):

``` coffeescript
util = require 'util'
Ector = require 'ector'
FileConceptNetwork =  require('file-concept-network').FileConceptNetwork

file_backup = "cn.json"
ector = new Ector()
ector.injectConceptNetwork FileConceptNetwork
ector.cn.load file_backup
previousResponseNodes = null
quiet = false
speakReplies = 
  [ 'Thanks.', 'OK', 'As you will.', 'Glad to be here again :)', ':)', 'Thank you.']

module.exports = (robot) ->

  ector.setName robot.name

  robot.respond /shut up/i, (msg) ->
    quiet = true

  robot.respond /speak/i, (msg) ->
    quiet = false
    msg.reply msg.random speakReplies

  robot.respond /save yourself/i, (msg) ->
    ector.cn.save file_backup
    msg.reply "Thanks, My mind is safe now!"

  robot.catchAll (msg) ->
    if not quiet
      text = msg.message.text
      ector.setUser msg.message.user.name
      ector.addEntry text
      ector.linkNodesToLastSentence previousResponseNodes
      response = ector.generateResponse()
      previousResponseNodes = response.nodes
      msg.reply response.sentence

```

**Comment by parmentf** [comment]
Now, I know how to add callbacks, here is the updated code (and I published a version 1.2.1):

``` coffeescript
util = require 'util'
Ector = require 'ector'
FileConceptNetwork =  require('file-concept-network').FileConceptNetwork

file_backup = "cn.json"
ector = new Ector()
ector.injectConceptNetwork FileConceptNetwork
ector.cn.load file_backup, (err) ->
  if err
    console.error 'Error while loading cn.json\n%s', err
previousResponseNodes = null
quiet = false
speakReplies = 
  [ 'Thanks.', 'OK', 'As you will.', 'Glad to be here again :)', ':)', 'Thank you.']

module.exports = (robot) ->

  ector.setName robot.name

  robot.respond /shut up/i, (msg) ->
    quiet = true

  robot.respond /speak/i, (msg) ->
    quiet = false
    msg.reply msg.random speakReplies

  robot.respond /save yourself/i, (msg) ->
    ector.cn.save file_backup, (err) ->
      if err
        msg.reply "An error occurred while saving cn.json:", err
      else
        msg.reply "Thanks, My mind is safe now!"

  robot.catchAll (msg) ->
    if not quiet
      text = msg.message.text
      ector.setUser msg.message.user.name
      ector.addEntry text
      ector.linkNodesToLastSentence previousResponseNodes
      response = ector.generateResponse()
      previousResponseNodes = response.nodes
      msg.reply response.sentence
```

**Comment by victorperin** [comment]
Oh, I really liked your chatbot, I and for me was a good way to learn.
So, thanks to you!
I'll update my bot and if i know a way to improve it, I will.
